src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryDispatchMessageInspector.cs

```diff
@@ -0,0 +1,146 @@
+﻿// <copyright file="TelemetryDispatchMessageInspector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#if NETFRAMEWORK
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IDispatchMessageInspector"/> implementation which adds telemetry to incoming requests.
+    /// </summary>
+    public class TelemetryDispatchMessageInspector : IDispatchMessageInspector
+    {
+        private static readonly ConcurrentDictionary<string, ActionMetadata> ActionMetadataCache = new ConcurrentDictionary<string, ActionMetadata>();
```

Comment:
Same thing about performance of ConcurrentDictionary. Hashtable with write-lock is probably better here.